# COS60008 Introduction to Data Science
Finish rest of the header

## Task 1 - Data Loading & Preparation

### Load three data files

In [1]:
# importing libraries
import pandas as pd
from pprint import pprint
# from IPython.core.interactiveshell import InteractiveShell
# InteractiveShell.ast_node_interactivity = "all"


In [2]:
# Load data
data_1_env = pd.read_csv("data_1_env.csv")
data_2_env = pd.read_csv("data_2_env.csv")
data_3_crop = pd.read_csv("data_3_crop.csv")

In [3]:
## Major steps
# 1. converting units 
# 2. duplicate rows
# 3. missing values
# 4. out of range values
# 5. inconsistent string values(labels)

### Explore data
* head() = to understand basic structure of data
* describe() = to understand statastical properies of data

In [4]:
print("data_1_env")
display(data_1_env.head())
display(data_1_env.describe())
print("data_2_env")
display(data_2_env.head())
display(data_2_env.describe())
print("data_3_crop")
display(data_3_crop.head())
display(data_3_crop.describe())

data_1_env


Year          Location  Avg Min Temp *F  Yr Rain inches Heat Wave   
0  1980  South Janethaven            68.18         31.5303         N  \
1  1980        Jesushaven            73.22         26.9331         N   
2  1980         Lake Eric            69.80         45.4181         N   
3  1981  South Janethaven            68.00         32.8792         N   
4  1981        Jesushaven            74.12         25.1740         N   

  Dry Spell Cold Wave Wet Spell  
0         N         Y         N  
1         N         Y         N  
2         N         N         N  
3         N         Y         N  
4         N         y         N

Year  Avg Min Temp *F  Yr Rain inches
count    21.00000        21.000000       21.000000
mean   1983.00000        70.682857       36.975386
std       2.04939         2.254607       10.269588
min    1980.00000        68.000000       25.075600
25%    1981.00000        68.360000       28.369300
50%    1983.00000        70.160000       33.704300
75%    1985.00000        73.220000       47.537800
max    1986.00000        74.120000       55.882300

data_2_env


Year          Location  Avg Min Temp *C  Yr Rain mm Heat Wave Dry Spell   
0  1987         Lake Eric             21.3     1482.06         N         N  \
1  1987  South Janethaven             20.3      946.64         N        No   
2  1987        Jesushaven             23.0      686.09         N         N   
3  1988         Lake Eric             21.1     1346.78         N         N   
4  1988  South Janethaven             20.3     1017.00         N         N   

  Cold Wave Wet Spell  
0         N         N  
1         Y         N  
2         Y         N  
3         N         N  
4         Y         N

Year  Avg Min Temp *C   Yr Rain mm
count    99.000000        99.000000    99.000000
mean   2003.000000        22.562626  1299.626364
std       9.570362         1.318826   352.451699
min    1987.000000        20.300000   686.090000
25%    1995.000000        21.550000  1026.740000
50%    2003.000000        22.300000  1276.830000
75%    2011.000000        23.800000  1570.740000
max    2019.000000        25.100000  2096.540000

data_3_crop


Year          Location Irrigation Crop Type  Crop Damage    Yield   
0  1980.0  South Janethaven          Y    Canola          1.0   913.34  \
1  1980.0        Jesushaven          Y    Canola          1.0  1022.55   
2  1980.0         Lake Eric         No      Oats          0.0  1196.80   
3  1981.0  South Janethaven          Y    Canola          1.0   906.17   
4  1981.0        Jesushaven          Y     Wheat          1.0  1735.82   

          Observer  
0      Jeremy Mann  
1       Joel Brown  
2  Connor Griffith  
3   Jennifer Adams  
4       Joel Brown

Year  Crop Damage        Yield
count   124.000000   124.000000   124.000000
mean   1999.500000     0.282258  1327.345645
std      11.722746     0.451924   312.137616
min    1980.000000     0.000000   775.220000
25%    1989.000000     0.000000  1046.897500
50%    1999.500000     0.000000  1402.700000
75%    2010.000000     1.000000  1618.297500
max    2019.000000     1.000000  1773.590000

* get unique values from each column to see data entry errors

In [5]:
def get_uniques(df):
    u = []
    for col in df:
        u.append(f"{col} ({str(type(df[col][0]))}): {df[col].unique()}")
    display(u)
print("data_1_env")
get_uniques(data_1_env)
print("data_2_env")
get_uniques(data_2_env)
print("data_3_crop")
get_uniques(data_3_crop)

data_1_env


["Year (<class 'numpy.int64'>): [1980 1981 1982 1983 1984 1985 1986]",
 "Location (<class 'str'>): ['South Janethaven' 'Jesushaven' 'Lake Eric']",
 "Avg Min Temp *F (<class 'numpy.float64'>): [68.18 73.22 69.8  68.   74.12 69.44 68.36 73.76 70.16 70.52 72.86 70.7\n 68.72 73.94]",
 "Yr Rain inches (<class 'numpy.float64'>): [31.5303 26.9331 45.4181 32.8792 25.174  48.4465 33.7043 25.0756 51.4508\n 40.1977 26.1681 47.5378 29.7473 28.3693 49.4835 28.378  55.8823 37.9721\n 50.3193 36.0292 25.7866]",
 "Heat Wave (<class 'str'>): ['N' '0']",
 "Dry Spell (<class 'str'>): ['N']",
 "Cold Wave (<class 'str'>): ['Y' 'N' 'y' '1']",
 "Wet Spell (<class 'str'>): ['N']"]

data_2_env


["Year (<class 'numpy.int64'>): [1987 1988 1989 1990 1991 1992 1993 1994 1995 1996 1997 1998 1999 2000\n 2001 2002 2003 2004 2005 2006 2007 2008 2009 2010 2011 2012 2013 2014\n 2015 2016 2017 2018 2019]",
 "Location (<class 'str'>): ['Lake Eric' 'South Janethaven' 'Jesushaven']",
 "Avg Min Temp *C (<class 'numpy.float64'>): [21.3 20.3 23.  21.1 23.6 21.4 20.8 23.5 21.7 23.4 20.7 23.9 23.3 21.5\n 20.5 20.6 24.  21.6 22.  24.2 24.1 21.9 23.7 22.2 21.  24.3 22.1 22.3\n 24.5 21.2 22.4 24.4 22.7 24.7 22.5 25.  24.6 22.9 24.8 22.8 24.9 25.1\n 23.2]",
 "Yr Rain mm (<class 'numpy.float64'>): [1482.06  946.64  686.09 1346.78 1017.    807.9  1183.95  938.16  826.31\n 1435.25  799.03 1063.17  790.66 1395.8  1032.08  748.65 1488.71  934.64\n  810.34 1285.51  961.96  791.51 1511.28 1038.37  785.36 1572.9  1252.89\n 1147.96  885.44 1706.17 1190.96  862.   1688.45 1139.7   865.91 1529.2\n  769.11 1611.12 1276.83 1556.08  854.42 1364.5  1584.31  850.26 1274.86\n 1285.19 1722.95  878.29  861.66 1358.3 

data_3_crop


["Year (<class 'numpy.float64'>): [1980. 1981. 1982. 1983. 1984. 1985. 1986. 1987. 1988. 1989. 1990. 1991.\n 1992. 1993. 1994. 1995. 1996. 1997. 1998. 1999. 2000. 2001. 2002. 2003.\n 2004. 2005. 2006. 2007. 2008. 2009. 2010.   nan 2011. 2012. 2013. 2014.\n 2015. 2016. 2017. 2018. 2019.]",
 "Location (<class 'str'>): ['South Janethaven' 'Jesushaven' 'Lake Eric' nan]",
 "Irrigation (<class 'str'>): ['Y' 'No' 'N' 'Yes' nan 'no']",
 "Crop Type (<class 'str'>): ['Canola' 'Oats' 'Wheat' nan]",
 "Crop Damage (<class 'numpy.float64'>): [ 1.  0. nan]",
 "Yield (<class 'numpy.float64'>): [ 913.34 1022.55 1196.8   906.17 1735.82 1497.96  917.47     nan 1431.4\n 1195.77  915.49 1421.54 1193.72 1611.36 1716.79  794.13 1732.13 1493.45\n 1616.42 1495.61 1324.48 1734.79  794.66 1323.2  1023.71 1196.33 1623.93\n 1442.87 1194.01 1642.56 1038.43 1191.48 1035.61  940.42 1749.49 1493.72\n  940.28 1034.09 1492.87 1630.24 1448.47 1193.45 1334.1  1052.3   792.63\n 1641.37 1737.03  788.49  944.07 1341.09 1442.

In [6]:
# fix data_1_env 

# convert Avg Min Temp from Ferenheit to Celsius
data_1_env.insert(2, "Avg Min Temp *C", (data_1_env['Avg Min Temp *F'] - 32) / 1.8)
data_1_env = data_1_env.drop("Avg Min Temp *F", axis=1)

# convert Yr Rain inches to int to remove string issues 
data_1_env.insert(3, "Yr Rain mm", data_1_env['Yr Rain inches'] * 25.4)
data_1_env = data_1_env.drop("Yr Rain inches", axis=1)



In [7]:
# universel replacemnet dict
replacement = {'No': 'N','no': 'N', '0': 'N', 0.: 'N', 'n': 'N', 'Yes': 'Y','yes': 'Y', '1': 'Y', 1.: 'Y', 'y' : 'Y'}
# Column to perform replacement on 
replacement_columns = ['Heat Wave', 'Dry Spell', 'Cold Wave', 'Wet Spell']
# Perform replacement on data_1_env
for column in replacement_columns:
    data_1_env[column] = data_1_env[column].replace(replacement)
# Perform replacement on data_2_env
for column in replacement_columns:
    data_2_env[column] = data_2_env[column].replace(replacement)


In [8]:
# first deal with nan values
print("Table 1: Rows with NaN values")
display(data_3_crop[data_3_crop.isnull().any(axis=1)])
print("Table 2: Duplicate rows.")
display(data_3_crop[data_3_crop.duplicated(subset=['Year', 'Location'], keep=False)])
print("Table 3: Rows where Michael Jimenez is observer.")
display(data_3_crop[data_3_crop['Observer'] == "Michael Jimenez"])
print("Table 4: Rows where location is South Janethaven")
display(data_3_crop[(data_3_crop['Location'] == "South Janethaven")&((data_3_crop['Year']>=2008)|(data_3_crop['Year'].isnull()))])

Table 1: Rows with NaN values


Year          Location Irrigation Crop Type  Crop Damage    Yield   
7    1982.0        Jesushaven          Y      Oats          1.0      NaN  \
23   1987.0  South Janethaven          Y      Oats          NaN  1323.20   
95      NaN  South Janethaven        Yes      Oats          0.0  1387.65   
107  2014.0         Lake Eric        NaN       NaN          1.0   775.22   
112  2015.0               NaN          Y     Wheat          0.0  1701.71   

              Observer  
7           Joel Brown  
23       James Johnson  
95                 NaN  
107  Jonathan Williams  
112    Michael Jimenez

Table 2: Duplicate rows.


Year          Location Irrigation Crop Type  Crop Damage    Yield   
7    1982.0        Jesushaven          Y      Oats          1.0      NaN  \
8    1982.0        Jesushaven          Y      Oats          1.0  1431.40   
23   1987.0  South Janethaven          Y      Oats          NaN  1323.20   
24   1987.0  South Janethaven          Y      Oats          1.0  1323.20   
106  2014.0         Lake Eric          N    Canola          1.0   775.22   
107  2014.0         Lake Eric        NaN       NaN          1.0   775.22   

              Observer  
7           Joel Brown  
8           Joel Brown  
23       James Johnson  
24       James Johnson  
106  Jonathan Williams  
107  Jonathan Williams

Table 3: Rows where Michael Jimenez is observer.


Year          Location Irrigation Crop Type  Crop Damage    Yield   
108  2014.0  South Janethaven          Y     Wheat          0.0  1700.02  \
111  2015.0  South Janethaven          Y     Wheat          0.0  1701.71   
112  2015.0               NaN          Y     Wheat          0.0  1701.71   
113  2016.0  South Janethaven          Y      Oats          0.0  1401.62   

            Observer  
108  Michael Jimenez  
111  Michael Jimenez  
112  Michael Jimenez  
113  Michael Jimenez

Table 4: Rows where location is South Janethaven


Year          Location Irrigation Crop Type  Crop Damage    Yield   
88   2008.0  South Janethaven          Y    Canola          0.0   977.00  \
91   2009.0  South Janethaven          Y     Wheat          0.0  1680.56   
94   2010.0  South Janethaven        Yes      Oats          0.0  1387.65   
95      NaN  South Janethaven        Yes      Oats          0.0  1387.65   
98   2011.0  South Janethaven          Y    Canola          0.0   979.60   
99   2012.0  South Janethaven          Y     Wheat          0.0  1709.16   
104  2013.0  South Janethaven          Y     Wheat          0.0  1695.69   
108  2014.0  South Janethaven          Y     Wheat          0.0  1700.02   
111  2015.0  South Janethaven          Y     Wheat          0.0  1701.71   
113  2016.0  South Janethaven          Y      Oats          0.0  1401.62   
116  2017.0  South Janethaven          Y    Canola          0.0  1000.00   
120  2018.0  South Janethaven          Y    Canola          0.0  1000.10   
123  2019.0  South Janethaven          Y      Oats          0.0  1403.78   

            Observer  
88    Charles Greene  
91    Charles Greene  
94    Charles Greene  
95               NaN  
98    Charles Greene  
99         Kara Hill  
104      Frank Sloan  
108  Michael Jimenez  
111  Michael Jimenez  
113  Michael Jimenez  
116    Theresa Jones  
120    Marilyn Watts  
123      Jose Jensen

* based on above Table 1 adn Table 2 we can see that there is duplicate rows for each row with NaN values. This means we can remove rows 7, 23, and 107
* based on above Table 3 we can see that Observer Michael Jimenez is only assigned to Location South Janethaven. we can only see, in Table 3 that if Loaction for row 112 replaced with South Janethaven, It is duplicate of row 111. We can remove row 112
* based on above Table 4 we can see that there's trend of having only 1 observation per year. For year 2010 and 2011 we have observations yet there is row 95 with NaN year. based on values from other columns, we can infer that row 95 is duplicate of row 94. We can remove row 95
* Above proves that there exist duplicate rows for all the NaN values thuse we can remove NaN rows.

In [9]:
# remove rows with NaN values from data_3_crops
data_3_crop = data_3_crop.dropna()

In [10]:
# Year from float to int 
data_3_crop['Year'] = data_3_crop['Year'].astype(int)
# universel replacemnet dict
replacement = {'no': 'N', 'No': 'N', 0.: 'N', 'Yes': 'Y', 1.: 'Y'}
# Column to perform replacement on 
replacement_columns = ['Irrigation', 'Crop Damage']
# Perform replacement 
for column in replacement_columns:
    data_3_crop[column] = data_3_crop[column].replace(replacement)

### Merge dfs 

In [14]:
len(data_1_env),len(data_2_env),len(data_3_crop)

(21, 99, 120)

In [17]:
data_1_env.columns,data_2_env.columns,data_3_crop.columns

(Index(['Year', 'Location', 'Avg Min Temp *C', 'Yr Rain mm', 'Heat Wave',
        'Dry Spell', 'Cold Wave', 'Wet Spell'],
       dtype='object'),
 Index(['Year', 'Location', 'Avg Min Temp *C', 'Yr Rain mm', 'Heat Wave',
        'Dry Spell', 'Cold Wave', 'Wet Spell'],
       dtype='object'),
 Index(['Year', 'Location', 'Irrigation', 'Crop Type', 'Crop Damage', 'Yield',
        'Observer'],
       dtype='object'))

In [23]:
data_1_2_env = pd.concat([data_1_env,data_2_env],ignore_index=True)
merged_data = data_1_2_env.merge(data_3_crop, on=['Year', 'Location'])

In [25]:
merged_data.head()

Year          Location  Avg Min Temp *C  Yr Rain mm Heat Wave Dry Spell   
0  1980  South Janethaven             20.1   800.86962         N         N  \
1  1980        Jesushaven             22.9   684.10074         N         N   
2  1980         Lake Eric             21.0  1153.61974         N         N   
3  1981  South Janethaven             20.0   835.13168         N         N   
4  1981        Jesushaven             23.4   639.41960         N         N   

  Cold Wave Wet Spell Irrigation Crop Type Crop Damage    Yield   
0         Y         N          Y    Canola           Y   913.34  \
1         Y         N          Y    Canola           Y  1022.55   
2         N         N          N      Oats           N  1196.80   
3         Y         N          Y    Canola           Y   906.17   
4         Y         N          Y     Wheat           Y  1735.82   

          Observer  
0      Jeremy Mann  
1       Joel Brown  
2  Connor Griffith  
3   Jennifer Adams  
4       Joel Brown

In [26]:
def get_uniques(df):
    u = []
    for col in df:
        u.append(f"{col} ({str(type(df[col][0]))}): {df[col].unique()}")
    display(u)
get_uniques(merged_data)

["Year (<class 'numpy.int64'>): [1980 1981 1982 1983 1984 1985 1986 1987 1988 1989 1990 1991 1992 1993\n 1994 1995 1996 1997 1998 1999 2000 2001 2002 2003 2004 2005 2006 2007\n 2008 2009 2010 2011 2012 2013 2014 2015 2016 2017 2018 2019]",
 "Location (<class 'str'>): ['South Janethaven' 'Jesushaven' 'Lake Eric']",
 "Avg Min Temp *C (<class 'numpy.float64'>): [20.1 22.9 21.  20.  23.4 20.8 20.2 23.2 21.2 21.4 22.7 21.5 20.4 23.3\n 21.3 20.3 23.  21.1 23.6 20.8 23.5 21.7 23.4 20.7 23.9 23.3 20.5 20.6\n 24.  21.6 22.  24.2 24.1 21.9 23.7 22.2 21.  24.3 22.1 22.3 24.5 22.4\n 24.4 24.7 22.5 25.  24.6 24.8 22.8 24.9 25.1 23.2]",
 "Yr Rain mm (<class 'numpy.float64'>): [ 800.86962  684.10074 1153.61974  835.13168  639.4196  1230.5411\n  856.08922  636.92024 1306.85032 1021.02158  664.66974 1207.46012\n  755.58142  720.58022 1256.8809   720.8012  1419.41042  964.49134\n 1278.11022  915.14168  654.97964 1482.06     946.64     686.09\n 1346.78    1017.       807.9     1183.95     938.16     826.

In [29]:
merged_data.to_csv('clean_data.csv', index=False)